### DISCUSSION TITLE
binary switch target value seems to ignore targetValue

### DISCUSSION THREAD

**Discussion by TheKorn2** [zwave-js/zwave-js-ui#1222] [Q&A] [answered]
I will preface this by saying it's entirely possible that I don't have the correct syntax.

I'm trying to get mqtt to issue switch binary on/off commands, testing with mosquitto_pub right now. Long story short, no matter what the value I pass in targetValue, the device turns on.  Pass it true, turns on.  Pass it false, turns on. Pass it a null, turns on. Pass it a string, turns on.

Seems basic, so I'm thinking I'm passing something badly and it's just doing its best guess?

```
$ mosquitto_pub -h (mqtt host) -u (mqtt username) -P (mqtt password) -t zwavemqtttest/_CLIENTS/ZWAVE_GATEWAY-zwavemqtttest/api/
sendCommand/set -m '{ "args":[{"nodeId":2,"commandClass":37,"endpoint":0},"set",[{"targetValue":false}]]}'
```

Which causes things to happen (ON!), and I get the following output...

```
2021-05-14 19:08:54.503 INFO MQTT: Message received on zwavemqtttest/_CLIENTS/ZWAVE_GATEWAY-zwavemqtttest/api/sendCommand/set, '{ "args":[{"nodeId":2,"commandClass":37,"endpoint":0},"set",[{"targetValue":false}]]}'
2021-05-14 19:08:54.505 INFO ZWAVE: Calling api sendCommand with args: [
  { nodeId: 2, commandClass: 37, endpoint: 0 },
  'set',
  [ { targetValue: false }, [length]: 1 ],
  [length]: 3
]
2021-05-14 19:08:54.553 INFO ZWAVE: Success zwave api call sendCommand { success: true, message: 'Success zwave api call', result: undefined }
```

So, I'm clearly not doing something right, so uh... how do I get it to turn off instead? :D Thanks!

**Comment by robertsLando** [comment, marked as answer]
Wait you are tring to write a value right? you shouldn't use sendCommand for this: https://zwave-js.github.io/zwavejs2mqtt/#/guide/mqtt?id=set-values

In poor words you should add a `/set` suffix to the values you want to write sending as payload the value you want to set

**Reply by TheKorn2** [reply]
@robertsLando No, I'm trying to turn it on/off using a binary switch command. I'm not trying to modify any of the configuration values on the device. (It doesn't have any AFAIK.)

**Reply by robertsLando** [reply]
> binary switch command

I know it could be a command but you can do the same using the way I tell you... Can you control that value from the UI control panel? If so you can also do it by using mqtt as the set sent from control panel is the same sent by using mqtt wiith `/set` suffix

**Reply by TheKorn2** [reply]
Mmmm, well that doesn't work either.  In fact that works worse; that way I can't turn it on nor turn it off.

Here's the topic if I manipulate it via the gui:

2021-05-17T16:22:19-0500 zwavemqtttest/2/37/0/currentValue:{"time":1621286539146,"value":true}
2021-05-17T16:22:20-0500 zwavemqtttest/2/37/0/currentValue:{"time":1621286540189,"value":true}

(((device turns on)))

And if I manipulate that topic directly...

2021-05-17T16:22:46-0500 zwavemqtttest/2/37/0/currentValue:{"time":1621286540199,"value":false}

(((nothing happens)))


2021-05-17T16:23:03-0500 zwavemqtttest/2/37/0/currentValue:{"time":1621286540199,"value":true}

(((still nothing happens)))


2021-05-17T16:23:05-0500 zwavemqtttest/2/37/0/currentValue:{"time":1621286540199,"value":false}

(((and yeah, nothing happens)))


So if I can't use the API to do a binary switch set command, and I can't manipulate the mqtt topic directly, how am I supposed to do it?

**Reply by robertsLando** [reply]
Publish that payload to `zwavemqtttest/2/37/0/targetValue/set`

**Reply by TheKorn2** [reply]
@robertsLando That got me fixed up, thanks!

**Comment by a-lurker** [comment]
As per the OP:   "I will preface this by saying it's entirely possible that I don't have the correct syntax." Plus I'm new to this app. My query seems similar to the OP's.

I am testing an old Aeon dimmer. I can control it through the UI with no problems. However if I do this:

```
zwave/_CLIENTS/ZWAVE_GATEWAY-zwave-js-ui/api/sendCommand/set
{
  "args": [
    {
      "nodeId": 4,
      "commandClass": 38,
      "endpoint": 0
    },
    "set",
    [
      {
        "targetValue":55
      }
    ]
  ]
}
```

I get:

```
2024-03-05 17:08:09.319 INFO Z-WAVE: Calling api sendCommand with args: [
  { nodeId: 4, commandClass: 38, endpoint: 0 },
  'set',
  [ { targetValue: 55 }, [length]: 1 ],
  [length]: 3
]
2024-03-05 17:08:09.322 INFO Z-WAVE: targetValue is not a number (ZW0322) sendCommand undefined
2024-03-05 17:08:09.325 INFO MQTT: Message received on zwave/_CLIENTS/ZWAVE_GATEWAY-zwave-js-ui/api/sendCommand: '{"success":false,"message":"targetValue is not a number (ZW0322)","args":[{"nodeId":4,"commandClass":38,"endpoint":0},"set",[{"targetValue":55}]],"origin":{"args":[{"nodeId":4,"commandClass":38,"endpoint":0},"set",[{"targetValue":55}]]}}'
```

If I change the argument name from "targetValue" to "Heidi" in the command, I still get the same result "targetValue is not a number". No matter what I do with the argument(s) they are totally ignored. Except if I leave out the array brackets around [{"targetValue":55}] then I get:

```
2024-03-05 17:23:37.529 INFO Z-WAVE: if args is given, it must be an array sendCommand undefined
```

So yes, I could use any one of the following (and I would like to) but none of them show anything in the log whatsoever and don't work. Noting I have adjusted the Gateway settings to accommodate the different variations:

```
zwave/Lounge_downstairs/Test_dimmer/switch_multilevel/endpoint_0/targetValue/set"

zwave/Test_dimmer/switch_multilevel/endpoint_0/targetValue/set"

zwave/4/38/0/targetValue/set"

With payload of 55 or {"value" : 55}
```

Just doesn't respond, although they show up in mqttexplorer OK. 

Note, if I do a get (which is successful):

```
zwave/_CLIENTS/ZWAVE_GATEWAY-zwave-js-ui/api/sendCommand/set
{
  "args": [
    {
      "nodeId": 4,
      "commandClass": 38,
      "endpoint": 0
    },
    "get"
  ]
}
```

The log says:

```
2024-03-05 17:41:10.771 INFO Z-WAVE: [Node 004] Value updated: 38-0-currentValue 55 => 55
17:41:10.772 DRIVER « [Node 004] [REQ] [ApplicationCommand]
                      └─[MultilevelSwitchCCReport]
                          current value: 55
2024-03-05 17:41:10.773 INFO Z-WAVE: Success zwave api call sendCommand { currentValue: 55, targetValue: undefined, duration: undefined }
```

Note that:   "targetValue: undefined". Shouldn't the result show "targetValue: 55"?

Any ideas? Thanks in advance.

**Reply by TheKorn2** [reply]
Well I just re-worked my old example (man, haven't used mosquitto_pub in forever!), and here are the results...

NO:
```
mosquitto_pub -h (mqtt host) -u (mqtt_username) -P (mqtt_pass) -t "zwave/73/38/0/targetValue/set" -m [{"value":"99"}] 
```

NO:
``` 
mosquitto_pub -h (mqtt host) -u (mqtt_username) -P (mqtt_pass) -t "zwave/73/38/0/targetValue/set" -m [{"99"}]
```

YES:
```
mosquitto_pub -h (mqtt host) -u (mqtt_username) -P (mqtt_pass) -t "zwave/73/38/0/targetValue/set" -m 99
```

So you're possibly over-thinking it (like I tend to do). Try sending just the raw value to targetValue/set, not as an array or quoted or anything.

**Reply by kpine** [reply]
Per the MQTT API documentation, [`sendCommand`](https://zwave-js.github.io/zwave-js-ui/#/guide/mqtt?id=sendcommand) is for invoking the CommandClass API directly, not using the values you see in the control panel.

The Multilevel Switch CC API [set](https://zwave-js.github.io/node-zwave-js/#/api/CCs/MultilevelSwitch?id=set) method does not take an object as an argument, it takes two positional arguments, target value and duration (optional).

```text
zwave/_CLIENTS/ZWAVE_GATEWAY-zwave-js-ui/api/sendCommand/set
{
  "args": [
    {
      "nodeId": 4,
      "commandClass": 38,
      "endpoint": 0
    },
    "set",
    [
      55
    ]
  ]
}
```

Using the CC API isn't usually recommended unless you have a special need.

**Reply by robertsLando** [reply]
Thanks @kpine you beat me by some seconds 😆 @TheKorn2 Like @kpine suggested the preferred way to write values is to use the valueId topic with the `/set` suffix like explained [here](https://zwave-js.github.io/zwave-js-ui/#/guide/mqtt?id=set-values)

**Reply by TheKorn2** [reply]
@robertsLando   @kpine   ???????  I am using /set !  (now)

Three years ago I wasn't, but that was three years ago.

**Reply by robertsLando** [reply]
@TheKorn2 sorry I misread the name in the post of 2h ago, I meant to tag @a-lurker as he asked this:

> I am testing an old Aeon dimmer. I can control it through the UI with no problems. However if I do this:
[...]
> If I change the argument name from "targetValue" to "Heidi" in the command, I still get the same result "targetValue is not a number". No matter what I do with the argument(s) they are totally ignored. Except if I leave out the array brackets around [{"targetValue":55}] then I get:
[...]

And them were referencing to the `sendCommand`

Anyway in your examples later you wrote `-m [{"value":"99"}]`, the accepted payloads are json or raw values so the valid ones would be:

`{"value":"99"}` (without square brackets) or just `99`. Note that in case of json you could also specify options like duration, see [here](https://zwave-js.github.io/zwave-js-ui/#/guide/mqtt?id=set-with-options)

**Reply by TheKorn2** [reply]
I was wondering why [{"value":"99"}] didn't work! Thanks!

**Comment by a-lurker** [comment]
Mmmm - tried that as per above:

With payload of 55 or {"value" : 55}

and I thought it would work. May be a secondary problem in my case. Still want to see why the api call doesn't work as advertised!

**Reply by robertsLando** [reply]
@a-lurker what's the topic you used to send that command?

**Reply by a-lurker** [reply]
This fixed the CC API call - works perfectly - thank you @kpine: "does not take an object as an argument, it takes two positional arguments, target value and duration (optional)."

```
zwave/_CLIENTS/ZWAVE_GATEWAY-zwave-js-ui/api/sendCommand/set
{
  "args": [
    {
      "nodeId": 4,
      "commandClass": 38,
      "endpoint": 0
    },
    "set",
    [
      55
    ]
  ]
}
```

However, @robertsLando I can't get any of these to work, which is really odd:

```
zwave/Lounge_downstairs/Test_dimmer/switch_multilevel/endpoint_0/targetValue/set

zwave/Test_dimmer/switch_multilevel/endpoint_0/targetValue/set

zwave/4/38/0/targetValue/set

With payload of 55 or {"value" : 55}
```
Nothing appears in the log but does in mqttexplorer.

**Reply by robertsLando** [reply]
> Nothing appears in the log but does in mqttexplorer.

Firstly be sure you are running the latest version. Then be sure you don't have a typo

**Reply by kpine** [reply]
Does it work if you use `{"value":55}` instead (no spaces)?
